### ISSUE TITLE
Invalid generated typescript with JsonSubTypes and v2.17.558

### ISSUE THREAD

**Issue by jdussouillez** [vojtechhabarta/typescript-generator#428]
The typescript generated with the [v2.17.558](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.17.558) is invalid when the `JsonSubTypes` annotation is present.

The code below worked with [v2.16.538](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.16.538) but not with [v2.17.558](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.17.558). The only way to generate valid TS using the below example is to remove the `JsonSubTypes` annotation on the `Entity` class.

### Entity.java

```java
package com.test.tsgen;

import com.fasterxml.jackson.annotation.JsonGetter;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY)
@JsonSubTypes({
    @JsonSubTypes.Type(value = Foo.class, name = "Foo"),
    @JsonSubTypes.Type(value = Bar.class, name = "Bar")
})
public abstract class Entity<T> {

    @JsonProperty
    protected final T id;

    public Entity(T id) {
        this.id = id;
    }

    public T getId() {
        return id;
    }

    /**
     * Renvoie le type de l'entité courante
     *
     * @return le type de l'entité courante
     */
    @JsonGetter("entityType")
    protected abstract EntityType getEntityType();
}
```

### Foo.java
```java
package com.test.tsgen;

public class Foo extends Entity<String> {

    public Foo(final String id) {
        super(id);
    }

    @Override
    protected EntityType getEntityType() {
        return EntityType.FOO;
    }
}
```

### Bar.java
```java
package com.test.tsgen;

public class Bar extends Entity<Integer> {

    public Bar(final int id) {
        super(id);
    }

    @Override
    protected EntityType getEntityType() {
        return EntityType.BAR;
    }
}
```

### EntityCollection.java
```java
package com.test.tsgen;

import java.util.Set;
import java.util.TreeSet;

public class EntityCollection {

    protected final Set<Entity<?>> entities = new TreeSet<>();

    public Set<Entity<?>> getEntities() {
        return entities;
    }
}
```

### pom.xml

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>com.test</groupId>
    <artifactId>typescript-generator-test</artifactId>
    <version>1.0.0-SNAPSHOT</version>
    <packaging>jar</packaging>
    <properties>
        <project.name>typescript-generator-test</project.name>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
    </properties>

    <dependencies>
        <!-- Jackson -->
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-annotations</artifactId>
            <version>2.10.0</version>
        </dependency>
    </dependencies>

    <build>
        <plugins>
            <plugin>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.8.1</version>
                <configuration>
                    <source>11</source>
                    <target>11</target>
                    <compilerArgument>-Xlint:all</compilerArgument>
                </configuration>
            </plugin>

            <plugin>
                <groupId>cz.habarta.typescript-generator</groupId>
                <artifactId>typescript-generator-maven-plugin</artifactId>
                <version>2.17.558</version>
                <executions>
                    <execution>
                        <id>generate</id>
                        <phase>process-classes</phase>
                        <goals>
                            <goal>generate</goal>
                        </goals>
                    </execution>
                </executions>
                <configuration>
                    <jsonLibrary>jackson2</jsonLibrary>
                    <classPatterns>
                        <classPattern>com.test.tsgen.**</classPattern>
                    </classPatterns>
                    <excludeClasses>
                        <excludeClass>java.io.Serializable</excludeClass>
                        <excludeClass>java.lang.Comparable</excludeClass>
                    </excludeClasses>
                    <mapClasses>asClasses</mapClasses>
                    <mapDate>asDate</mapDate>
                    <mapEnum>asEnum</mapEnum>
                    <nonConstEnums>true</nonConstEnums>
                    <mapPackagesToNamespaces>true</mapPackagesToNamespaces>
                    <optionalProperties>useLibraryDefinition</optionalProperties>
                    <outputFile>model.ts</outputFile>
                    <outputFileType>implementationFile</outputFileType>
                    <outputKind>module</outputKind>
                </configuration>
            </plugin>
        </plugins>
    </build>
</project>
```

### v2.16.538

```typescript
/* tslint:disable */
/* eslint-disable */
// Generated using typescript-generator version 2.16.538 on 2019-10-23 16:06:54.

export namespace com.test.tsgen {
    export class Entity<T> {
        "@type": "Bar" | "Foo";
        id?: T;
        entityType?: com.test.tsgen.EntityType;
    }
}

export namespace com.test.tsgen {
    export class Bar extends com.test.tsgen.Entity<number> {
        "@type": "Bar";
    }
}

export namespace com.test.tsgen {
    export class EntityCollection {
        entities?: com.test.tsgen.Entity<any>[];
    }
}

export namespace com.test.tsgen {
    export class Foo extends com.test.tsgen.Entity<string> {
        "@type": "Foo";
    }
}

export namespace com.test.tsgen {
    export enum EntityType {
        FOO = "FOO",
        BAR = "BAR",
    }
}
```

**This output is valid**

### v2.17.558

```typescript
/* tslint:disable */
/* eslint-disable */
// Generated using typescript-generator version 2.17.558 on 2019-10-23 16:07:59.

export namespace com.test.tsgen {
    export class Entity<T> {
        "@type": "Bar" | "Foo";
        id?: T;
        entityType?: com.test.tsgen.EntityType;
    }
}

export namespace com.test.tsgen {
    export class Bar extends com.test.tsgen.Entity<number> {
        "@type": "Bar";
        id?: number;
    }
}

export namespace com.test.tsgen {
    export class EntityCollection {
        entities?: com.test.tsgen.EntityUnion<any>[];
    }
}

export namespace com.test.tsgen {
    export class Foo extends com.test.tsgen.Entity<string> {
        "@type": "Foo";
        id?: string;
    }
}

export namespace com.test.tsgen {
    export type EntityUnion<T> = com.test.tsgen.Foo<> | com.test.tsgen.Bar<>; // <-- ISSUE IS HERE
}

export namespace com.test.tsgen {
    export enum EntityType {
        FOO = "FOO",
        BAR = "BAR",
    }
}
```

**This output is not valid**.

```console
$ tsc model.ts
model.ts(43,34): error TS2315: Type 'Foo' is not generic.
model.ts(43,52): error TS1099: Type argument list cannot be empty.
model.ts(43,57): error TS2315: Type 'Bar' is not generic.
model.ts(43,75): error TS1099: Type argument list cannot be empty.
```

**Comment by jdussouillez** [comment]
Project to reproduce the bug (branch `issues/428`) : https://github.com/jdussouillez/typescript-generator-test/tree/issues/428

**Comment by vojtechhabarta** [comment]
Thanks for precise bug report.
I committed fix in d02b6494507fcd0b70d35e8892d33501714f1990.
Could you please try it if it works now?

**Comment by jdussouillez** [comment]
@vojtechhabarta It works ! :tada: 

Thank you for the quick fix.

Output with [v2.17.564](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.17.564) :

```typescript
/* tslint:disable */
/* eslint-disable */
// Generated using typescript-generator version 2.17-SNAPSHOT on 2019-10-24 08:19:04.

export namespace com.test.tsgen {

    export class Entity<T> {
        "@type": "Bar" | "Foo";
        id?: T;
        entityType?: com.test.tsgen.EntityType;
    }

}

export namespace com.test.tsgen {

    export class Bar extends com.test.tsgen.Entity<number> {
        "@type": "Bar";
        id?: number;
    }

}

export namespace com.test.tsgen {

    export class EntityCollection {
        entities?: com.test.tsgen.EntityUnion<any>[];
    }

}

export namespace com.test.tsgen {

    export class Foo extends com.test.tsgen.Entity<string> {
        "@type": "Foo";
        id?: string;
    }

}

export namespace com.test.tsgen {

    export type EntityUnion<T> = com.test.tsgen.Foo | com.test.tsgen.Bar;

}

export namespace com.test.tsgen {

    export enum EntityType {
        FOO = "FOO",
        BAR = "BAR",
    }

}
```

**Comment by danielwegener** [comment]
Arr - is there any chance to get this fix released to maven central? :) started to migrate from spring-fox today and instantly got bitten by this issue

**Comment by vojtechhabarta** [comment]
Released in version [2.18.565](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.18.565).
